nixos/modules/config/nix-channel.nix

```diff
@@ -94,10 +95,31 @@ in
       NIX_PATH = cfg.nixPath;
     };
 
-    nix.settings.nix-path = mkIf (! cfg.channel.enable) (mkDefault "");
-
     systemd.tmpfiles.rules = lib.mkIf cfg.channel.enable [
       ''f /root/.nix-channels - - - - ${config.system.defaultChannel} nixos\n''
     ];
+
+    system.activationScripts.no-nix-channel = mkIf (!cfg.channel.enable)
+      (stringAfter [ "etc" "users" ] ''
+        explainChannelWarning=0
+        if [ -e "/root/.nix-defexpr/channels" ]; then
+            warn '/root/.nix-defexpr/channels exists, but channels have been disabled.'
+            explainChannelWarning=1
+        fi
+        if [ -e "/nix/var/nix/profiles/per-user/root/channels" ]; then
+            warn "/nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled."
+            explainChannelWarning=1
+        fi
+        getent passwd | while IFS=: read -r _ _ _ _ _ home _ ; do
+            if [ -n  "$home" -a -e "$home/.nix-defexpr/channels" ]; then
+                warn "$home/.nix-defexpr/channels exists, but channels have been disabled." 1>&2
+                explainChannelWarning=1
```

Comment:
This is set in a subshell so it will not be visible from outside the loop.
Flipping the redirect to be `while ... done < <(getent passwd)`, or using `shopt -s lastpipe` (bash-ism) should fix it.

Say, shouldn't ShellCheck have flagged this?

Reply:
ohh..
I was only able to shellcheck a small part.
It'd be good to shellcheck the whole activation script, but that's a much larger exercise

Reply:
I'll put this string in a file. Let's see

Reply:
Shellcheck caught it too, but thank you for spotting this!